File: test/error-payload.test.ts
// Regression gate for the Python worker's _error_payload classifier: managed
// refusals (empty wallet, spent instance budget) must surface as
// quota_exhausted from their message text alone. Pure unit tests — no live
// Hermes/LLM needed. The assertions live in test/error_payload_test.py.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { spawnSync } from 'node:child_process';

test('hermes worker _error_payload classifies managed refusals as quota_exhausted', () => {
  const result = spawnSync('python3', ['test/error_payload_test.py'], { encoding: 'utf8' });
  assert.equal(result.status, 0, `python unit tests failed:\n${result.stdout}\n${result.stderr}`);
});

File: test/error_payload_test.py
#!/usr/bin/env python3
"""Regression tests for _error_payload: the failure-text classifier that turns a
provider/agent exception into the worker's {code, message, hint} payload.

Hermes summarizes an HTTP 402 body down to its `message` (the `type` field is
dropped), so both managed refusals must classify by their message text alone.
"""

import sys
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "server" / "workers"))

import hermes_worker


class ErrorPayloadTest(unittest.TestCase):
    def test_wallet_empty_refusal_is_quota_exhausted(self):
        payload = hermes_worker._error_payload(
            Exception("HTTP 402: Workspace balance exhausted. Top up the workspace wallet to continue.")
        )
        self.assertEqual(payload["code"], "quota_exhausted")
        self.assertIn("workspace balance", payload["hint"])

    def test_instance_budget_refusal_is_quota_exhausted(self):
        payload = hermes_worker._error_payload(
            Exception("HTTP 402: Instance budget exhausted. Raise the monthly cap or top up this instance to continue.")
        )
        self.assertEqual(payload["code"], "quota_exhausted")

    def test_retry_wrapped_budget_refusal_beats_rate_limit(self):
        payload = hermes_worker._error_payload(
            Exception("Rate limited after 3 retries — HTTP 402: Instance budget exhausted.")
        )
        self.assertEqual(payload["code"], "quota_exhausted")

    def test_unrelated_failure_stays_generic(self):
        payload = hermes_worker._error_payload(Exception("upstream connect error"))
        self.assertEqual(payload["code"], "worker_error")
        self.assertNotIn("hint", payload)


if __name__ == "__main__":
    unittest.main()
